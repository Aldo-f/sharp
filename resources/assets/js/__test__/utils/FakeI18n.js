export default {
    install() {
        window.i18n = {};
    }
}