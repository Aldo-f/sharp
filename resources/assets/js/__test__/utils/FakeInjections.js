import Vue from 'vue';

const injectedComponents = ['$field', '$form'];

export default {
    provide() {
        return injectedComponents.reduce((injections, prop) => {
            injections[prop] = new Vue();
            return injections;
        }, {});
    }
}