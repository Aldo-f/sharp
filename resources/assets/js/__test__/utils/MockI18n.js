import Localization ,* as localizationFn from '../../mixins/Localization';

const mockLangImplementation = localeKey => `{{ ${localeKey} }}`;

export default {
    install() {
        window.i18n = {};
    },

    mockLangFunction() {
        localizationFn.lang = jest.fn(mockLangImplementation);
        Localization.methods.l = jest.fn(mockLangImplementation);
    }
}