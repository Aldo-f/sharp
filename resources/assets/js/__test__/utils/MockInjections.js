import Vue from 'vue';


const injectedStrings = {
    xsrfToken: 'xsrfTest'
};


const injectedComponents = {
    $field:{},
    $form:{
        data: () => ({
            errors: {}
        })
    },
    $tab:{},

    actionsBus:{}
};

function resolveComponents(comps) {
    return Object.keys(comps).reduce((injections, compName) => {
        let Comp = Vue.extend(comps[compName]);
        injections[compName] = new Comp();
        return injections;
    }, {});
}

export default {
    provide() {
        return {
            ...resolveComponents(injectedComponents),
            ...injectedStrings
        };
    }
}