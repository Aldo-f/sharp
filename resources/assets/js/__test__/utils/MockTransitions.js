export default {
    install() {
        window.getComputedStyle = jest.fn(() => ({
            animationDelay:'',
            animationDuration:'',
            transitionDelay:'',
            transitionDuration:''
        }));
    }
}