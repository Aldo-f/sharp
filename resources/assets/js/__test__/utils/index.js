export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC } from './mockSFC';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}