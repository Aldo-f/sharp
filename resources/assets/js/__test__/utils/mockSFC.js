export function mockSFC (sfcModule, { template, render, ...extend }={}, mockOptions={}) {

    let { isJsFile } = mockOptions;

    if(isJsFile) {
        sfcModule.__oldDefault__ = { ...sfcModule.default }
    }

    Object.assign(sfcModule.default, {
        mixins: [],
        'extends': extend,
        template: template,
        render: !template ? render || function() { return this._v(`__MOCKED_SFC_${sfcModule.name}__\n`) } : null,
        watch: {},
        beforeCreate: ()=>{},
        created: ()=>{},
        beforeMount: ()=>{},
        mounted: ()=>{},
        beforeUpdate: ()=>{},
        updated: ()=>{},
        beforeDestroy: ()=>{},
        destroyed: ()=>{},
        data: ()=>({}),
        provide: {},
        inject: [],
    });

    //console.log(sfcModule);
    return sfcModule.default;
}

export function unmockSFC(sfcModule) {
    if(sfcModule.__oldDefault__) {
        sfcModule.default = sfcModule.__oldDefault__;
        return;
    }

    let { 'default':defaultObj, ...options } = sfcModule;
    sfcModule.default = options;
}