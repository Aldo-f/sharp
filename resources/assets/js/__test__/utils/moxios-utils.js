import moxios from 'moxios';

export function nextRequestFulfilled(config) {
    return new Promise(resolve => {
        moxios.wait(async () => {
            let request = moxios.requests.mostRecent();
            let response = await request.respondWith(config);
            resolve(response); // moxios Response object
        })
    });
}
