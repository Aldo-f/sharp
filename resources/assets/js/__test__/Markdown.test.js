import Vue from 'vue/dist/vue';
import Markdown from '../components/form/fields/markdown/Markdown.vue';

import { MockI18n, MockInjections } from './utils';

import * as Localization from '../mixins/Localization';



describe('markdown-field', () => {

    Vue.component('sharp-markdown', Markdown);
    Vue.use(MockI18n);

    beforeEach(()=>{
        document.body.innerHTML = `
            <div id="app">
                <sharp-markdown :value="value" 
                    :read-only="readOnly" 
                    placeholder="Champ md" 
                    :toolbar="toolbar" 
                    :height="310"
                    :max-image-size="3"
                    :locale="locale"
                    @input="inputEmitted">
                </sharp-markdown>
            </div>
        `;
        // mock range functions
        document.body.createTextRange = () => ({
            getBoundingClientRect: () => ({ }),
            getClientRects: () => ({ })
        });
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    it('can mount Markdown field', async () => {
        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can mount "localized" Markdown field', async () => {
        Localization.lang = jest.fn(() => 'LOCALIZED');
        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can mount "read only" Markdown field', async () => {
        await createVm({
            propsData: {
                readOnly: true
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('update value on locale changed', async () => {
        let $markdown = await createVm({
            propsData: {
                locale: 'fr'
            },
            data: ()=>({
                value: { text:'Valeur 1' }
            })
        });

        let { $root:vm, simplemde } = $markdown;

        vm.value.text = 'Valeur 2';
        vm.locale = 'en';

        await Vue.nextTick();

        expect(simplemde.value()).toBe('Valeur 2');
    });

    it('expose appropriate props to simplemde', async () => {
        let $markdown = await createVm({
            propsData: {
                toolbar: [{ name:'my action' }]
            },
            data: ()=>({
                value: { text:'Valeur 1' }
            })
        });

        let { simplemde } = $markdown;
        let { textarea } = $markdown.$refs;

        expect(simplemde.options).toMatchObject({
            element: textarea,
            initialValue: 'Valeur 1',
            placeholder: 'Champ md',
            spellChecker: false,
            autoDownloadFontAwesome: false,
            toolbar: [{ name:'my action' }],
        });


    });

    it('bound toolbar buttons custom action properly', async () =>{
        let $markdown = await createVm({
            propsData: {
                toolbar: [{ name:'image'}]
            }
        });

        let { simplemde } = $markdown;
        $markdown.insertUploadImage = jest.fn();

        expect(simplemde.toolbar[0].action).toBeInstanceOf(Function);
        simplemde.toolbar[0].action();

        expect($markdown.insertUploadImage).toHaveBeenCalled();
    });

    it('set read only properly', async () => {
        let $markdown = await createVm();

        let { simplemde } = $markdown;
        let { codemirror } = simplemde;

        expect(codemirror.getOption('readOnly')).toBe(false);

        $markdown.setReadOnly();

        expect(codemirror.getOption('readOnly')).toBe(true);
    });

    it('add codemirror event listener properly', async () => {
        let $markdown = await createVm();

        let { simplemde: {codemirror}} = $markdown;

        let callback = jest.fn();
        codemirror.on = jest.fn();

        $markdown.codemirrorOn('event', callback);
        expect(codemirror.on).toHaveBeenCalledWith('event', callback);
        expect(callback).not.toHaveBeenCalled();

        $markdown.codemirrorOn('event', callback, true);
        expect(codemirror.on).toHaveBeenCalledTimes(2);
        expect(codemirror.on).toHaveBeenLastCalledWith('event', callback);
        expect(callback).toHaveBeenCalled();
    });

    it('emit input on text changed', async () => {
        let inputEmitted = jest.fn();
        let $markdown = await createVm({
            methods: {
                inputEmitted
            }
        });

        let { simplemde } = $markdown;

        simplemde.value('AAA');

        expect(inputEmitted).toHaveBeenLastCalledWith(expect.objectContaining({ text:'AAA' }));
    });

    it('insert image properly', async () => {
        let $markdown = await createVm();

        let { simplemde } = $markdown;
        let { codemirror } = simplemde;

        codemirror.getSelection = jest.fn(() => 'Image title');
        codemirror.markText = jest.fn(() => ({
            on: jest.fn()
        }));

        $markdown.insertUploadImage(simplemde);

        console.log(simplemde.value());
        //console.log($markdown.$root._provided);
    });
});

async function createVm(customOptions={}) {

    const vm = new Vue({
        el: '#app',
        mixins: [customOptions, MockInjections],

        props:['readOnly', 'toolbar', 'locale'],

        'extends': {
            data:() => ({
                value: {}
            }),
            methods: {
                inputEmitted: ()=>{}
            }
        }
    });

    await Vue.nextTick();

    return vm.$children[0];
}