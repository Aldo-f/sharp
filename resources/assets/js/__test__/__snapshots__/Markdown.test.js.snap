// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown-field can mount "localized" Markdown field 1`] = `
<div id="app">
    <div class="SharpMarkdown">
        <div class="SharpModule__inner"><textarea style="display: none;"></textarea>
            <div class="editor-toolbar"><a title="LOCALIZED (Ctrl-B)" tabindex="-1" class="fa fa-bold"></a><a title="LOCALIZED (Ctrl-I)" tabindex="-1" class="fa fa-italic"></a><a title="LOCALIZED (Ctrl-H)" tabindex="-1" class="fa fa-header"></a><i class="separator">|</i><a title="LOCALIZED (Ctrl-')"
                    tabindex="-1" class="fa fa-quote-left"></a><a title="LOCALIZED (Ctrl-L)" tabindex="-1" class="fa fa-list-ul"></a><a title="LOCALIZED (Ctrl-Alt-L)" tabindex="-1" class="fa fa-list-ol"></a><i class="separator">|</i><a title="LOCALIZED (Ctrl-K)"
                    tabindex="-1" class="fa fa-link"></a><a title="LOCALIZED" tabindex="-1" class="fa fa-picture-o"></a><i class="separator">|</i><a title="LOCALIZED (Ctrl-P)" tabindex="-1" class="fa fa-eye no-disable"></a><a title="LOCALIZED (F9)" tabindex="-1"
                    class="fa fa-columns no-disable no-mobile"></a><a title="LOCALIZED (F11)" tabindex="-1" class="fa fa-arrows-alt no-disable no-mobile"></a><i class="separator">|</i><a title="LOCALIZED" tabindex="-1" class="fa fa-question-circle" href="https://simplemde.com/markdown-guide"
                    target="_blank"></a></div>
            <div class="CodeMirror cm-s-paper CodeMirror-wrap CodeMirror-empty" style="width: auto; height: 310px;">
                <div style="overflow: hidden; position: relative; width: 3px; height: 0px;"><textarea style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;" autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0"></textarea></div>
                <div class="CodeMirror-vscrollbar" cm-not-content="true">
                    <div style="min-width: 1px;"></div>
                </div>
                <div class="CodeMirror-hscrollbar" cm-not-content="true">
                    <div style="height: 100%; min-height: 1px;"></div>
                </div>
                <div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div>
                <div class="CodeMirror-gutter-filler" cm-not-content="true"></div>
                <div class="CodeMirror-scroll" tabindex="-1">
                    <div class="CodeMirror-sizer" style="margin-left: 0px;">
                        <div style="position: relative;">
                            <div class="CodeMirror-lines" role="presentation">
                                <div style="position: relative; outline: none;" role="presentation">
                                    <pre style="height: 0px; overflow: visible;" class="CodeMirror-placeholder">Champ md</pre>
                                    <div class="CodeMirror-measure">
                                        <pre><span>xxxxxxxxxx</span></pre>
                                    </div>
                                    <div class="CodeMirror-measure"></div>
                                    <div style="position: relative; z-index: 1;"></div>
                                    <div class="CodeMirror-cursors"></div>
                                    <div class="CodeMirror-code" role="presentation"></div>
                                </div>
                            </div>
                        </div>
                    </div>
                    <div style="position: absolute; height: 30px; width: 1px;"></div>
                    <div class="CodeMirror-gutters" style="display: none;"></div>
                </div>
            </div>
            <div class="editor-preview-side"></div>
            <div class="editor-statusbar"><span class="autosave"></span><span class="lines">1</span><span class="words">0</span><span class="cursor">0:0</span></div>
        </div>
    </div>
</div>
`;

exports[`markdown-field can mount "read only" Markdown field 1`] = `
<div id="app">
    <div class="SharpMarkdown SharpMarkdown--read-only">
        <div class="SharpModule__inner"><textarea style="display: none;"></textarea>
            <div class="editor-toolbar"><a tabindex="-1" class="fa fa-bold"></a><a tabindex="-1" class="fa fa-italic"></a><a tabindex="-1" class="fa fa-header"></a><i class="separator">|</i><a tabindex="-1" class="fa fa-quote-left"></a><a tabindex="-1" class="fa fa-list-ul"></a>
                <a
                    tabindex="-1" class="fa fa-list-ol"></a><i class="separator">|</i><a tabindex="-1" class="fa fa-link"></a><a tabindex="-1" class="fa fa-picture-o"></a><i class="separator">|</i><a tabindex="-1" class="fa fa-eye no-disable"></a><a tabindex="-1" class="fa fa-columns no-disable no-mobile"></a>
                    <a
                        tabindex="-1" class="fa fa-arrows-alt no-disable no-mobile"></a><i class="separator">|</i><a tabindex="-1" class="fa fa-question-circle"></a></div>
            <div class="CodeMirror cm-s-paper CodeMirror-wrap CodeMirror-empty" style="width: auto; height: 310px;">
                <div style="overflow: hidden; position: relative; width: 3px; height: 0px;"><textarea style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;" autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0"></textarea></div>
                <div class="CodeMirror-vscrollbar" cm-not-content="true">
                    <div style="min-width: 1px;"></div>
                </div>
                <div class="CodeMirror-hscrollbar" cm-not-content="true">
                    <div style="height: 100%; min-height: 1px;"></div>
                </div>
                <div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div>
                <div class="CodeMirror-gutter-filler" cm-not-content="true"></div>
                <div class="CodeMirror-scroll" tabindex="-1">
                    <div class="CodeMirror-sizer" style="margin-left: 0px;">
                        <div style="position: relative;">
                            <div class="CodeMirror-lines" role="presentation">
                                <div style="position: relative; outline: none;" role="presentation">
                                    <pre style="height: 0px; overflow: visible;" class="CodeMirror-placeholder">Champ md</pre>
                                    <div class="CodeMirror-measure">
                                        <pre><span>xxxxxxxxxx</span></pre>
                                    </div>
                                    <div class="CodeMirror-measure"></div>
                                    <div style="position: relative; z-index: 1;"></div>
                                    <div class="CodeMirror-cursors"></div>
                                    <div class="CodeMirror-code" role="presentation"></div>
                                </div>
                            </div>
                        </div>
                    </div>
                    <div style="position: absolute; height: 30px; width: 1px;"></div>
                    <div class="CodeMirror-gutters" style="display: none;"></div>
                </div>
            </div>
            <div class="editor-preview-side"></div>
            <div class="editor-statusbar"><span class="autosave"></span><span class="lines">1</span><span class="words">0</span><span class="cursor">0:0</span></div>
        </div>
    </div>
</div>
`;

exports[`markdown-field can mount Markdown field 1`] = `
<div id="app">
    <div class="SharpMarkdown">
        <div class="SharpModule__inner"><textarea style="display: none;"></textarea>
            <div class="editor-toolbar"><a tabindex="-1" class="fa fa-bold"></a><a tabindex="-1" class="fa fa-italic"></a><a tabindex="-1" class="fa fa-header"></a><i class="separator">|</i><a tabindex="-1" class="fa fa-quote-left"></a><a tabindex="-1" class="fa fa-list-ul"></a>
                <a
                    tabindex="-1" class="fa fa-list-ol"></a><i class="separator">|</i><a tabindex="-1" class="fa fa-link"></a><a tabindex="-1" class="fa fa-picture-o"></a><i class="separator">|</i><a tabindex="-1" class="fa fa-eye no-disable"></a><a tabindex="-1" class="fa fa-columns no-disable no-mobile"></a>
                    <a
                        tabindex="-1" class="fa fa-arrows-alt no-disable no-mobile"></a><i class="separator">|</i><a tabindex="-1" class="fa fa-question-circle" href="https://simplemde.com/markdown-guide" target="_blank"></a></div>
            <div class="CodeMirror cm-s-paper CodeMirror-wrap CodeMirror-empty" style="width: auto; height: 310px;">
                <div style="overflow: hidden; position: relative; width: 3px; height: 0px;"><textarea style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;" autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0"></textarea></div>
                <div class="CodeMirror-vscrollbar" cm-not-content="true">
                    <div style="min-width: 1px;"></div>
                </div>
                <div class="CodeMirror-hscrollbar" cm-not-content="true">
                    <div style="height: 100%; min-height: 1px;"></div>
                </div>
                <div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div>
                <div class="CodeMirror-gutter-filler" cm-not-content="true"></div>
                <div class="CodeMirror-scroll" tabindex="-1">
                    <div class="CodeMirror-sizer" style="margin-left: 0px;">
                        <div style="position: relative;">
                            <div class="CodeMirror-lines" role="presentation">
                                <div style="position: relative; outline: none;" role="presentation">
                                    <pre style="height: 0px; overflow: visible;" class="CodeMirror-placeholder">Champ md</pre>
                                    <div class="CodeMirror-measure">
                                        <pre><span>xxxxxxxxxx</span></pre>
                                    </div>
                                    <div class="CodeMirror-measure"></div>
                                    <div style="position: relative; z-index: 1;"></div>
                                    <div class="CodeMirror-cursors"></div>
                                    <div class="CodeMirror-code" role="presentation"></div>
                                </div>
                            </div>
                        </div>
                    </div>
                    <div style="position: absolute; height: 30px; width: 1px;"></div>
                    <div class="CodeMirror-gutters" style="display: none;"></div>
                </div>
            </div>
            <div class="editor-preview-side"></div>
            <div class="editor-statusbar"><span class="autosave"></span><span class="lines">1</span><span class="words">0</span><span class="cursor">0:0</span></div>
        </div>
    </div>
</div>
`;
